File: ios/Packages/PadelCore/Sources/PadelCore/Model/GolpeVisible.swift
import Foundation

/// El repertorio que la app le **enseña** al jugador, que no es el mismo que el que
/// distingue el detector por dentro.
///
/// ### Por qué existe esta capa
///
/// El detector separa nueve tipos; la app enseña siete. La diferencia no es cosmética,
/// es una decisión medida. Con las dos tandas limpias de pista (82 golpes etiquetados,
/// ago 2026) el acierto es:
///
/// | | ocho tipos | repertorio visible |
/// |---|---|---|
/// | tanda de 42 | 71 % | **85 %** |
/// | tanda de 40 (calibrada) | 80 % | **82 %** |
/// | las dos juntas | 68 % | **79 %** |
///
/// Las dos distinciones que se pliegan son justo las dos que los datos dicen que hoy no
/// se pueden sostener:
///
/// - **Víbora dentro de bandeja.** Las dos tandas se contradicen: en una la bandeja se
///   golpea más alta que la víbora y en la otra más baja. Mientras un jugador no
///   calibre con sus propias tandas, separarlas es echar una moneda al aire con dos
///   nombres.
/// - **El lado de la volea.** Una volea es un bloqueo sin muñeca, y el efecto que
///   decidiría el lado no está en la señal: las diez voleas de la tanda de 42 midieron
///   entre 0,1 y 3,9 de rotación axial, que es ruido.
///
/// Nada de esto se pierde: el detector **sigue** produciendo los nueve tipos y las
/// tandas se graban con ellos, que es lo que alimenta al modelo entrenado. Esto solo
/// decide qué se le enseña a una persona.
///
/// Espejo de `GolpeVisible` en el core Kotlin, con los tests allí.
public enum GolpeVisible: String, CaseIterable, Sendable {
    case derecha
    case reves
    case volea
    case globo
    case bandeja
    case remate
    case saque

    public var etiqueta: String {
        switch self {
        case .derecha: return "Derecha"
        case .reves: return "Revés"
        case .volea: return "Volea"
        case .globo: return "Globo"
        case .bandeja: return "Bandeja"
        case .remate: return "Remate"
        case .saque: return "Saque"
        }
    }

    /// El golpe que se le enseña al jugador, o nil si no hay nada que enseñar.
    public static func de(_ tipo: ShotType) -> GolpeVisible? {
        switch tipo {
        case .forehand: return .derecha
        case .backhand: return .reves
        case .forehandVolley, .backhandVolley: return .volea
        case .lob: return .globo
        // La víbora se pliega dentro de la bandeja: las dos son el golpe alto de
        // control y hoy no se separan con garantías. Ver la cabecera.
        case .bandeja, .vibora: return .bandeja
        case .smash: return .remate
        case .serve: return .saque
        case .unknown: return nil
        }
    }

    /// Agrupa notas **promediando**, no sumando: una nota de 1 a 7 no se acumula. Si un
    /// jugador dio bandejas de 3,0 y víboras de 3,4, su bandeja visible es 3,2.
    public static func agruparNotas(_ porTipo: [ShotType: Float]) -> [(GolpeVisible, Float)] {
        var acumulado: [GolpeVisible: [Float]] = [:]
        for (tipo, nota) in porTipo {
            guard let visible = de(tipo) else { continue }
            acumulado[visible, default: []].append(nota)
        }
        // En el orden del repertorio, no en el del diccionario: la ficha de un partido
        // tiene que leerse igual siempre.
        return allCases.compactMap { visible in
            guard let notas = acumulado[visible], !notas.isEmpty else { return nil }
            return (visible, notas.reduce(0, +) / Float(notas.count))
        }
    }

    /// Agrupa recuentos **sumando**: doce bandejas y tres víboras son quince.
    public static func agruparRecuentos(_ porTipo: [ShotType: Int]) -> [(GolpeVisible, Int)] {
        var acumulado: [GolpeVisible: Int] = [:]
        for (tipo, cuantos) in porTipo where cuantos > 0 {
            guard let visible = de(tipo) else { continue }
            acumulado[visible, default: 0] += cuantos
        }
        return allCases.compactMap { visible in
            guard let total = acumulado[visible] else { return nil }
            return (visible, total)
        }
    }
}

File: android/core/src/main/kotlin/com/risingpadel/core/model/GolpeVisible.kt
package com.risingpadel.core.model

/**
 * El repertorio que la app le **enseña** al jugador, que no es el mismo que el que
 * distingue el detector por dentro.
 *
 * ### Por qué existe esta capa
 *
 * El detector separa diez tipos; la app enseña nueve. **La única que se pliega es la
 * víbora dentro de la bandeja**, y no por gusto: las dos tandas limpias de pista se
 * contradicen sobre cuál de las dos se golpea más alta, así que separarlas sin calibrar
 * es echar una moneda al aire con dos nombres puestos.
 *
 * Medido sobre los 82 golpes etiquetados de esas dos tandas (ago 2026):
 *
 * | | diez tipos | repertorio visible |
 * |---|---|---|
 * | tanda de 42 | 71 % | **76 %** |
 * | tanda de 40 (calibrada) | 80 % | **82 %** |
 * | las dos juntas | 68 % | **74 %** |
 *
 * ### Por qué la volea y el globo SÍ llevan lado
 *
 * Plegar también el lado de la volea daba más acierto de tabla —85 % en la tanda de 42
 * en vez de 76 %— y aun así se descartó: para un jugador no es lo mismo tener floja la
 * volea de derecha que la de revés, y una app que no se lo puede decir no le sirve para
 * entrenar. El número de la tabla mide otra cosa que lo útil que es el dato.
 *
 * Hay que decirlo claro de todas formas: **el lado de la volea hoy es poco fiable**. Lo
 * decide el signo de la rotación axial, y las diez voleas de la tanda de 42 midieron
 * entre 0,1 y 3,9 rad/s, que es ruido. En el globo el mismo signo tiene mejor pinta —es
 * un swing completo con muñeca, no un bloqueo— pero no hay tanda con la que decirlo.
 * Las dos cosas las arregla el modelo entrenado, no un umbral.
 *
 * El detector **sigue** produciendo los diez tipos y las tandas se graban con ellos, que
 * es lo que alimenta al modelo. Esto solo decide qué se le enseña a una persona: cuando
 * el modelo sepa separar víbora de bandeja de verdad, aquí se deshace el pliegue.
 */
enum class GolpeVisible(val etiqueta: String) {
    DERECHA("Derecha"),
    REVES("Revés"),
    VOLEA_DERECHA("Volea de derecha"),
    VOLEA_REVES("Volea de revés"),
    GLOBO_DERECHA("Globo de derecha"),
    GLOBO_REVES("Globo de revés"),
    BANDEJA("Bandeja"),
    REMATE("Remate"),
    SAQUE("Saque");

    companion object {
        /** El golpe que se le enseña al jugador, o null si no hay nada que enseñar. */
        fun de(tipo: ShotType): GolpeVisible? = when (tipo) {
            ShotType.FOREHAND -> DERECHA
            ShotType.BACKHAND -> REVES
            ShotType.FOREHAND_VOLLEY -> VOLEA_DERECHA
            ShotType.BACKHAND_VOLLEY -> VOLEA_REVES
            ShotType.FOREHAND_LOB -> GLOBO_DERECHA
            ShotType.BACKHAND_LOB -> GLOBO_REVES
            // La víbora se pliega dentro de la bandeja: las dos son el golpe alto de
            // control y hoy no se separan con garantías. Ver la cabecera.
            ShotType.BANDEJA, ShotType.VIBORA -> BANDEJA
            ShotType.SMASH -> REMATE
            ShotType.SERVE -> SAQUE
            ShotType.UNKNOWN -> null
        }

        /**
         * Agrupa notas por golpe visible **promediando**, no sumando: una nota de 1 a 7
         * no se acumula. Si un jugador dio bandejas de 3,0 y víboras de 3,4, su bandeja
         * visible es 3,2.
         *
         * El promedio es simple y no ponderado por cantidad a propósito: ponderar haría
         * que dos víboras sueltas movieran menos la nota que veinte bandejas, lo cual es
         * defendible, pero la nota por tipo ya viene promediada desde el estimador y
         * volver a ponderar aquí exigiría arrastrar los recuentos hasta este punto para
         * ganar una diferencia de centésimas.
         */
        fun agruparNotas(porTipo: Map<ShotType, Float>): Map<GolpeVisible, Float> {
            val acumulado = LinkedHashMap<GolpeVisible, MutableList<Float>>()
            for ((tipo, nota) in porTipo) {
                val visible = de(tipo) ?: continue
                acumulado.getOrPut(visible) { mutableListOf() }.add(nota)
            }
            return entries
                .mapNotNull { visible ->
                    acumulado[visible]?.let { visible to it.average().toFloat() }
                }
                .toMap()
        }

        /** Agrupa recuentos **sumando**: doce bandejas y tres víboras son quince. */
        fun agruparRecuentos(porTipo: Map<ShotType, Int>): Map<GolpeVisible, Int> {
            val acumulado = LinkedHashMap<GolpeVisible, Int>()
            for ((tipo, cuantos) in porTipo) {
                if (cuantos <= 0) continue
                val visible = de(tipo) ?: continue
                acumulado[visible] = (acumulado[visible] ?: 0) + cuantos
            }
            return entries.mapNotNull { v -> acumulado[v]?.let { v to it } }.toMap()
        }
    }
}
